File: src/grabs.Graphics.D3D11/D3D11Swapchain.cs
using System.Diagnostics.CodeAnalysis;
using grabs.Core;
using TerraFX.Interop.DirectX;
using TerraFX.Interop.Windows;
using static TerraFX.Interop.DirectX.DXGI_SWAP_EFFECT;
using static TerraFX.Interop.DirectX.DXGI;
using static TerraFX.Interop.Windows.Windows;

namespace grabs.Graphics.D3D11;

[SuppressMessage("Interoperability", "CA1416:Validate platform compatibility")]
internal sealed unsafe class D3D11Swapchain : Swapchain
{
    public override bool IsDisposed { get; protected set; }
    
    public override Format BufferFormat { get; }

    private readonly IDXGISwapChain* _swapchain;

    private D3D11Texture _swapchainTexture;

    public D3D11Swapchain(IDXGIFactory1* factory, ID3D11Device* device, ref readonly SwapchainInfo info)
    {
        BufferFormat = info.Format;
        
        DXGI_SWAP_CHAIN_DESC swapchainDesc = new()
        {
            OutputWindow = (HWND) ((D3D11Surface) info.Surface).Hwnd,
            Windowed = true,

            BufferDesc = new DXGI_MODE_DESC
            {
                Width = info.Size.Width,
                Height = info.Size.Height,
                Format = info.Format.ToD3D()
            },
            BufferCount = info.NumBuffers,
            BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT,

            SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD,
            SampleDesc = new DXGI_SAMPLE_DESC(1, 0),
        };

        GrabsLog.Log("Creating swapchain.");
        fixed (IDXGISwapChain** swapchain = &_swapchain)
            factory->CreateSwapChain((IUnknown*) device, &swapchainDesc, swapchain).Check("Create swapchain");

        GrabsLog.Log("Getting swapchain texture.");
        ID3D11Texture2D* swapchainTexture;
        _swapchain->GetBuffer(0, __uuidof<ID3D11Texture2D>(), (void**) &swapchainTexture)
            .Check("Get swapchain texture");

        _swapchainTexture = new D3D11Texture(device, swapchainTexture, info.Size);
    }
    
    public override Texture GetNextTexture()
    {
        return _swapchainTexture;
    }
    
    public override void Present()
    {
        _swapchain->Present(1, 0).Check("Present");
    }
    
    public override void Dispose()
    {
        if (IsDisposed)
            return;
        IsDisposed = true;

        _swapchainTexture.Dispose();
        _swapchain->Release();
    }
}

File: src/D3D11/D3D11Swapchain.cpp
#include "D3D11Swapchain.h"
#include "D3D11Utils.h"
#include "../Common.h"

namespace grabs::D3D11
{
    D3D11Swapchain::D3D11Swapchain(IDXGIFactory1* factory, ID3D11Device* device, DXGISurface* surface,
        const SwapchainDescription& description)
    {
        const DXGI_FORMAT format = Utils::FormatToD3D(description.Format);
        const UINT width = description.Size.Width;
        const UINT height = description.Size.Height;

        DXGI_SWAP_CHAIN_DESC desc
        {
            .BufferDesc = { .Width = width, .Height = height, .Format = format },
            .SampleDesc = { .Count = 1, .Quality = 0 },
            .BufferUsage = DXGI_USAGE_RENDER_TARGET_OUTPUT,
            .BufferCount = description.NumBuffers,
            .OutputWindow = surface->Window,
            .Windowed = TRUE,
            .SwapEffect = DXGI_SWAP_EFFECT_FLIP_DISCARD,
            .Flags = DXGI_SWAP_CHAIN_FLAG_ALLOW_TEARING | DXGI_SWAP_CHAIN_FLAG_ALLOW_MODE_SWITCH
        };

        D3D11_CHECK_RESULT(factory->CreateSwapChain(device, &desc, &Swapchain));

        ID3D11Texture2D* texture = nullptr;
        D3D11_CHECK_RESULT(Swapchain->GetBuffer(0, __uuidof(ID3D11Texture2D), reinterpret_cast<void**>(&texture)));

        ID3D11RenderTargetView* target = nullptr;
        D3D11_CHECK_RESULT(device->CreateRenderTargetView(texture, nullptr, &target));

        SwapchainTexture = std::make_unique<D3D11Texture>(texture, target);
    }

    D3D11Swapchain::~D3D11Swapchain()
    {
        Swapchain->Release();
    }

    Texture* D3D11Swapchain::GetNextTexture()
    {
        return SwapchainTexture.get();
    }

    void D3D11Swapchain::Present()
    {
        D3D11_CHECK_RESULT(Swapchain->Present(1, 0));
    }
}
